### PR TITLE
Fixing the interaction of signals in GAP and the IO package.

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This patch is part 1 of fixing the interaction of signals in GAP and the IO package.

This adds a new function (which I imagine might also be useful to anyone making GAP into a library) to pass information about dead child processes to GAP.

Once this is merged in, I have another PR for IO, which will call this function. Then I can remove the line in the test which forces IO's signal handler to be disabled (because then the test will still work even with IO's signal handler).

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
The IO part of this is at:

https://github.com/gap-packages/io/pull/55 (just for reference).

File: src/iostream.c [hunk 1]
```diff
@@ -593,14 +593,14 @@ Obj FuncCLOSE_PTY_IOSTREAM(Obj self, Obj stream)
...
     int status;
     int retcode = close(PtyIOStreams[pty].ptyFD);
     if (retcode)
         Pr("Strange close return code %d\n", retcode, 0);
     kill(PtyIOStreams[pty].childPID, SIGTERM);
+    // GAP (or another library) might wait on this PID before
+    // we handle it. If that happens, waitpid will return -1.
     retcode = waitpid(PtyIOStreams[pty].childPID, &status, WNOHANG);
     if (retcode == 0) {
         // Give process a second to quit
```
Comment (markuspf) [inline src/iostream.c]: I know this wasn't added in this PR (but the previous one which I didn't review), but the one second seems rather arbitrary (if usually enough for a process to quit). Does it make sense to have this waiting time configurable, or at least definable?

Comment (ChrisJefferson) [reply src/iostream.c]: I admit it's arbitary, but I'm not sure I'd want to add a configuration option people might not understand.

In general, if an app hasn't quite within a second of being sent a `SIGTERM`, then it's a fairly badly behaved program.

One option would be to wait less long before going back to see if the program has stopped, maybe waiting 0.1 seconds 10 times, so GAP can continue faster.

Comment (ChrisJefferson) [reply src/iostream.c]: Now you bring it up, I'm tempted to make the timeout much longer (say 10 seconds), and check more often. That should deal with almost every case in practice, and make sure GAP keeps working for particularly badly behaved subproblems.

We would never have found this if there wasn't buggy copies of `rev` floating around which swallow all signals and enter an infinite loop.

File: tst/teststandard/processes/children.tst [hunk 2]
```diff
@@ -0,0 +1,18 @@
+gap> d := DirectoryCurrent();;
+gap> scriptdir := DirectoriesLibrary( "tst/teststandard/processes/" );;
+gap> checkpl := Filename(scriptdir, "check.pl");;
+
+# If IO is loaded, disable it's signal handler
+gap> if IsBoundGlobal("IO_RestoreSIGCHLDHandler") then
+> ValueGlobal("IO_RestoreSIGCHLDHandler")();
```
Comment (fingolfin) [inline tst/teststandard/processes/children.tst]: I assume that's just for the current `io` version, until https://github.com/gap-packages/io/pull/55 is merged?

Comment (ChrisJefferson) [reply tst/teststandard/processes/children.tst]: Yes, if you comment this out, the test fails (it "should" fail, the manual explictly says that GAP's process stuff can't be used if IO's signal handler is enabled, it just got more likely to trigger).

File: src/iostream.c [hunk 3]
```diff
@@ -242,6 +242,30 @@ static UInt OpenPty(int * master, int * slave)
...
+{
+    GAP_ASSERT(childPID > 0);
+    GAP_ASSERT((WIFEXITED(status) || WIFSIGNALED(status)));
+    HashLock(PtyIOStreams);
+    for (UInt i = 0; i < MAX_PTYS; i++) {
+        if (PtyIOStreams[i].inuse && PtyIOStreams[i].childPID == childPID) {
+            PtyIOStreams[i].changed = 1;
+            PtyIOStreams[i].status = status;
+            PtyIOStreams[i].blocked = 0;
+            return 1;
```
Comment (fingolfin) [inline src/iostream.c]: This is missing an `HashUnlock(PtyIOStreams);`. Or perhaps use a single `return result;` at the end, and set `result = 1;` here, and init `ini result = 0;` at the top. Whatever.

**Comment by ChrisJefferson** [comment]
Hopefully all issues fixed.

File: tst/teststandard/processes/children.tst [hunk 4]
```diff
@@ -0,0 +1,18 @@
+gap> d := DirectoryCurrent();;
+gap> scriptdir := DirectoriesLibrary( "tst/teststandard/processes/" );;
+gap> checkpl := Filename(scriptdir, "check.pl");;
+
+# If IO is loaded, disable it's signal handler
```
Comment (fingolfin) [inline tst/teststandard/processes/children.tst]: "it's" -> "its"

**Pull request merged**